cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/PersistedMessageCallback.java

```diff
@@ -120,10 +112,7 @@ private void recordLastOffset() {
       if (lastOffset >= 0) {
         offsets.put(lastTopicPartition, lastOffset);
       }
-      if (messageCount.get() >= persistThreshold) {
-        messageCount.set(0);
-        persist();
-      }
+      persist();
```

Comment:
I think we shouldn't do it in the Iterator. Writes to metrics table are actually buffered for the whole Iterator (which is controlled by the fetch size when fetching from Kafka). We should only persist the offset after actually writing to metrics table, otherwise it's possible that we lost metrics, due to consumption of this Iterator is just to create Metrics record in memory.